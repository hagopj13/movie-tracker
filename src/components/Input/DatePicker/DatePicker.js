// @flow
import React from 'react';
import type { Moment } from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import ClearIcon from '@material-ui/icons/Clear';
import { KeyboardDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

type Props = {
  label: string,
  selectedDate: Moment,
  onDateChange: (date: Moment) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    marginLeft: theme.spacing(0.5),
  },
  label: {
    minWidth: theme.spacing(5),
    width: theme.spacing(5),
  },
  datePicker: {
    width: '100%',
  },
  icon: {
    padding: 0,
  },
}));

const DatePicker = (props: Props) => {
  const { label, selectedDate, onDateChange } = props;

  const classes = useStyles();

  const handleDateChange = (date: Moment) => {
    if (date !== selectedDate && !date?.isSame(selectedDate)) {
      onDateChange(date);
    }
  };

  return (
    <div className={classes.root}>
      <Typography className={classes.label}>{label}</Typography>
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <KeyboardDatePicker
          className={classes.datePicker}
          disableToolbar
          variant="inline"
          inputVariant="outlined"
          format="DD/MM/YYYY"
          margin="dense"
          color="primary"
          value={selectedDate}
          onChange={handleDateChange}
          InputProps={{
            endAdornment: (
              <IconButton className={classes.icon} onClick={() => handleDateChange(null)}>
                <ClearIcon />
              </IconButton>
            ),
          }}
          InputAdornmentProps={{
            position: 'start',
          }}
          KeyboardButtonProps={{
            className: classes.icon,
          }}
        />
      </MuiPickersUtilsProvider>
    </div>
  );
};

export default DatePicker;
