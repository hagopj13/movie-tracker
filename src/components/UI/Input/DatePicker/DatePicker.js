// @flow
import React from 'react';
import type { Moment } from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import ClearIcon from '@material-ui/icons/Clear';
import { KeyboardDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

type Props = {
  label: string,
  selectedDate: Moment,
  onDateChange: (date: Moment) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    marginLeft: theme.spacing(0.5),
  },
  label: {
    minWidth: 40,
    width: 40,
    fontWeight: 300,
  },
  icon: {
    padding: 0,
  },
}));

const DatePicker = (props: Props) => {
  const { label, selectedDate, onDateChange } = props;

  const classes = useStyles();

  return (
    <div className={classes.root}>
      {label && <Typography className={classes.label}>{label}</Typography>}
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <KeyboardDatePicker
          disableToolbar
          variant="inline"
          inputVariant="outlined"
          format="DD/MM/YYYY"
          margin="dense"
          color="primary"
          value={selectedDate}
          onChange={onDateChange}
          InputProps={{
            endAdornment: (
              <IconButton className={classes.icon} onClick={() => onDateChange(null)}>
                <ClearIcon />
              </IconButton>
            ),
          }}
          InputAdornmentProps={{
            position: 'start',
          }}
          KeyboardButtonProps={{
            className: classes.icon,
          }}
        />
      </MuiPickersUtilsProvider>
    </div>
  );
};

export default DatePicker;
