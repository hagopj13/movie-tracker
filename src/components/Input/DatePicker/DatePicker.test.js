import React from 'react';
import moment from 'moment';
import { createShallow } from '@material-ui/core/test-utils';
import { KeyboardDatePicker } from '@material-ui/pickers';
import Typography from '@material-ui/core/Typography';

import DatePicker from './DatePicker';

describe('DatePicker component', () => {
  let shallow;
  let label;
  let selectedDate;
  let wrapper;

  const mockDateChange = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    label = 'Some label';
    selectedDate = moment('2020-01-01');
    const mockProps = {
      label,
      selectedDate,
      onDateChange: mockDateChange,
    };
    wrapper = shallow(<DatePicker {...mockProps} />);
  });

  it('should render the DatePicker component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the label', () => {
    expect(wrapper.find(Typography).text()).toBe(label);
  });

  it('should pass the selectedDate prop as value to the KeyboardDatePicker', () => {
    expect(wrapper.find(KeyboardDatePicker).prop('value')).toBe(selectedDate);
  });

  it('should call the onDateChange prop with new date when the selected date changed', () => {
    const newDate = selectedDate.clone().add(1, 'day');
    wrapper.find(KeyboardDatePicker).prop('onChange')(newDate);
    expect(mockDateChange).toHaveBeenLastCalledWith(newDate);
  });
});
