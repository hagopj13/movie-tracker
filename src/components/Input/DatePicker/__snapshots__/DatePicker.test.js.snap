// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker component should render the DatePicker component correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-label-2"
  >
    Some label
  </WithStyles(ForwardRef(Typography))>
  <MuiPickersUtilsProvider
    utils={[Function]}
  >
    <PickerWithState
      InputAdornmentProps={
        Object {
          "position": "start",
        }
      }
      InputProps={
        Object {
          "endAdornment": <ForwardRef(WithStyles)
            className="makeStyles-icon-4"
            onClick={[Function]}
          >
            <UNDEFINED />
          </ForwardRef(WithStyles)>,
        }
      }
      KeyboardButtonProps={
        Object {
          "className": "makeStyles-icon-4",
        }
      }
      allowKeyboardControl={true}
      className="makeStyles-datePicker-3"
      color="primary"
      disableToolbar={true}
      format="DD/MM/YYYY"
      inputVariant="outlined"
      invalidDateMessage="Invalid Date Format"
      margin="dense"
      maxDate={2100-01-01T00:00:00.000Z}
      maxDateMessage="Date should not be after maximal date"
      minDate={1900-01-01T00:00:00.000Z}
      minDateMessage="Date should not be before minimal date"
      onChange={[Function]}
      openTo="date"
      value={"2019-12-31T23:00:00.000Z"}
      variant="inline"
      views={
        Array [
          "year",
          "date",
        ]
      }
    />
  </MuiPickersUtilsProvider>
</div>
`;
